clients/da-vinci-client/src/main/java/com/linkedin/davinci/validation/KafkaDataIntegrityValidator.java

```diff
@@ -55,28 +77,29 @@ public void validateMessage(
       PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long> consumerRecord,
       boolean endOfPushReceived,
       Lazy<Boolean> tolerateMissingMsgs) throws DataValidationException {
-    final GUID producerGUID = consumerRecord.getValue().producerMetadata.producerGUID;
-    ProducerTracker producerTracker = registerProducer(producerGUID);
-    producerTracker.validateMessage(consumerRecord, endOfPushReceived, tolerateMissingMsgs);
+    PartitionTracker partitionTracker = registerPartition(consumerRecord.getPartition());
+    partitionTracker.validateMessage(consumerRecord, endOfPushReceived, tolerateMissingMsgs);
   }
 
   /**
    * For a given partition, find all the producers that has written to this partition and update the offsetRecord using
-   * segment information.
-   * @param partition
-   * @param offsetRecord
+   * segment information. Prior to this, the state which is expired according to {@link #maxAgeInMs} will be cleared.
+   *
+   * @param partition to extract info for
+   * @param offsetRecord to modify
    */
   public void updateOffsetRecordForPartition(int partition, OffsetRecord offsetRecord) {
-    producerTrackerMap.values().forEach(producerTracker -> producerTracker.updateOffsetRecord(partition, offsetRecord));
+    PartitionTracker partitionTracker = registerPartition(partition);
+    if (this.maxAgeInMs != DISABLED) {
+      partitionTracker.clearExpiredStateAndUpdateOffsetRecord(offsetRecord, this.maxAgeInMs);
+    } else {
+      partitionTracker.updateOffsetRecord(offsetRecord);
+    }
   }
 
   public void cloneProducerStates(int partition, KafkaDataIntegrityValidator newValidator) {
-    for (Map.Entry<GUID, ProducerTracker> entry: producerTrackerMap.entrySet()) {
-      GUID producerGUID = entry.getKey();
-      ProducerTracker sourceProducerTracker = entry.getValue();
-      ProducerTracker destProducerTracker = newValidator.registerProducer(producerGUID);
-      sourceProducerTracker.cloneProducerStates(partition, destProducerTracker);
-    }
+    PartitionTracker destPartitionTracker = newValidator.registerPartition(partition);
```

Comment:
Leader DIV validator will clone the producer states from the regular DIV validator since the cleanup will happen when first restoring the regular DIV validator state, the producer state copied to Leader DIV validator should be minimal.
This issue can be addressed when we implement the Leader DIV validator periodical cleanup later.